pkg/queue/health/probe.go

```diff
@@ -17,18 +17,80 @@ limitations under the License.
 package health
 
 import (
+	"crypto/tls"
+	"errors"
+	"fmt"
 	"net"
+	"net/http"
 	"time"
+
+	corev1 "k8s.io/api/core/v1"
 )
 
+// HTTPProbeConfigOptions holds the HTTP probe config options
+type HTTPProbeConfigOptions struct {
+	Timeout time.Duration
+	*corev1.HTTPGetAction
+}
+
+// TCPProbeConfigOptions holds the TCP probe config options
+type TCPProbeConfigOptions struct {
+	SocketTimeout time.Duration
+	Address       string
+}
+
 // TCPProbe checks that a TCP socket to the address can be opened.
 // Did not reuse k8s.io/kubernetes/pkg/probe/tcp to not create a dependency
 // on klog.
-func TCPProbe(addr string, socketTimeout time.Duration) error {
-	conn, err := net.DialTimeout("tcp", addr, socketTimeout)
+func TCPProbe(config TCPProbeConfigOptions) error {
+	conn, err := net.DialTimeout("tcp", config.Address, config.SocketTimeout)
 	if err != nil {
 		return err
 	}
 	conn.Close()
 	return nil
 }
+
+// HTTPProbe checks that HTTP connection can be established to the address.
+func HTTPProbe(config HTTPProbeConfigOptions) error {
+	httpClient := &http.Client{
+		Transport: &http.Transport{
+			DisableKeepAlives: true,
+			TLSClientConfig: &tls.Config{
+				InsecureSkipVerify: true,
+			},
+		},
+		Timeout: config.Timeout,
+	}
+	url := fmt.Sprintf("%s://%s:%d%s", config.Scheme, config.Host, config.Port.IntValue(), config.Path)
+	req, err := http.NewRequest("GET", url, nil)
+	if err != nil {
+		return fmt.Errorf("Error constructing request %s", err.Error())
+	}
+
+	for _, header := range config.HTTPHeaders {
+		req.Header.Add(header.Name, header.Value)
+	}
+
+	var res *http.Response
+	res, err = httpClient.Do(req)
+	if err != nil {
+		return err
+	}
+
+	if !IsHTTPProbeReady(res) {
+		return errors.New("http probe did not respond Ready")
+	}
+
+	return nil
+}
+
+// IsHTTPProbeReady checks whether we received a successful Response
+func IsHTTPProbeReady(res *http.Response) bool {
```

Comment:
Any reason why this is public?

Reply:
We actually had another component which executed similar response code logic (more context check #4600 PR) and we thought having consistent check would be good idea so made public function. In next PR we should have changes from another component use this function.

Reply:
SG>